### ISSUE TITLE
Support on-update: edit-only

### ISSUE THREAD

**Issue by Xiphe** [exercism/pr-commenter-action#37]
_First of all thanks for providing and maintaining this great action!_

## Problem

I use this to communicate preview deployments to PRs. 
Since the preview is shut down once a PR is closed, I also update the comment to remove the now dead link.

There are scenarios where the creation comment can not be created and the PR is then closed.
This currently creates a "The preview has been closed" comment, while there never had been a preview.

I want to prevent creation of this comment and currently to me it seems presence of the creation comment would be the easiest way to determine the behavior.

## Proposed solution

For this it would be great if the action would support some kind of update-only behavior where nothing is done when no existing comment can be found. Also I'd need to switch the on-update behavior for the creation and deletion account.

My API proposal for this would be:

```yaml
# .github/pr-commenter.yml
comment:
  on-update: {{ onUpdate }}
```

```yaml
# .github/workflows/myFlow.yml
# ...
    steps:
      - ...shutdownPreview...
      - uses: exercism/pr-commenter-action@v1.3.0
        with:
          template-variables: |
            {
              "onUpdate": "edit-only",
            }
```

---

Happy to provide a PR if this sounds interesting to you. Also open for different API proposals. 
Cheers 🤙

**Comment by angelikatyborska** [comment]
Hi!

First of all, sorry for taking so long to reply 🙏 I was taking a break from OSS for a while.

I have to admit I don't fully understand your use case. I'm not sure what "preview deployment" or "preview comment" means 🤔 I'm going to base my reply on this suggestion:

> it would be great if the action would support some kind of update-only behavior where nothing is done when no existing comment can be found

I'm not a fan of this proposal because it adds a new problem: the action would need to somehow keep track of how many times it was successfully run on the same PR. Currently the logic behind the "on-update" option is simple - it only checks if another comment from this action already exists.

The only way I can think of for a GitHub action to persist information on a PR without adding a comment, iseither to add a PR label or edit the PR's description and add some HTML comments in there with metadata. Both seem pretty intrusive to me.

Do I understand correctly that on those PRs on which you no longer want the PR commenter to update comments, you always have to do some manual work? Edit the comment left by the PR commenter to remove some link? If that is the case, I think it could work for you to also label those PRs at that moment, and then configure the PR commenter action not to run on PRs that have this label ([I believe like this](https://github.com/orgs/community/discussions/26712#discussioncomment-3253012)).

**Comment by Xiphe** [comment]
Hi 👋 thanks a lot for the response and welcome back to OSS. 

Sorry that my use-cases might have been confusing, but I think you got the problem perfectly 👍 

> the action would need to somehow keep track of how many times it was successfully run on the same PR

Similar to how the "on-update" method I would have implemented the "edit-only" to check if another comment from this action exists. If yes => edit it; If no => do nothing.

> Do I understand correctly that on those PRs on which you no longer want the PR commenter to update comments

This is not what I'm aiming at. I want no new comment to be created when none was there in the first place. But if one was there I want it to be updated.

--

Diving into my example again: 

Currently when I open a PR on my project I deploy a preview version of the website
Once that is done pr-commenter-action comments "Hey here is the preview: example.org/preview-123" 
When the PR is closed the preview deployment is shut down
Once that is done pr-commenter-action edits the comment "Preview is shut down because PR was closed"

Now when for some reason the deployment fails
Then the "Hey here is the preview: example.org/preview-123" comment is not made (since there is no preview)
When I now close the PR still the "Preview is shut down because PR was closed" comment is created
Besides that the comment is incorrect, this creates a GitHub notification for people subscribed to the PR which I consider unnecessary noise. 

--

As I said, I'd happily provide a PR maybe that makes the case a little clearer. 
Thanks a lot for taking the time to discuss this 🤙

**Comment by angelikatyborska** [comment]
> Currently when I open a PR on my project I deploy a preview version of the website
> Once that is done pr-commenter-action comments "Hey here is the preview: example.org/preview-123"
> When the PR is closed the preview deployment is shut down
> Once that is done pr-commenter-action edits the comment "Preview is shut down because PR was closed"

Does this mean that deploying a preview version of the website somehow creates a new file in the PR? And shutting down the preview deployment does the same, but another file?

> I would have implemented the "edit-only" to check if another comment from this action exists. If yes => edit it; If no => do nothing.

> I want no new comment to be created when none was there in the first place.

In such a scenario, how can the comment ever exist? If the PR commenter is not supposed to create a comment if no previous comment exist?

**Comment by Xiphe** [comment]
Hi

No, the Creation of the website does not create files on the PR. 

> In such a scenario, how can the comment ever exist? If the PR commenter is not supposed to create a comment if no previous comment exists?

When the `on-update` setting is set through a template, the action can be triggered multiple times with different values. Let's assume this config file for the action:

```yaml
# .github/pr-commenter.yml
comment:
  # Note the template variable here, allowing different settings
  on-update: {{ onUpdate }}
  snippets:
    - id: preview-created-{{ id }}
      files:
        - "**/*"
      body: |
        {{#create}}
        ### 🛫 Preview deployed!

        Visit {{{ url }}} for a test flight.
        {{/create}}
        {{#shutdown}}
        ### 🛬 Preview removed

        The preview has been shut down since the PR was closed
        {{/shutdown}}
```

The Deploy action might look like this and is triggered on every change on a PR

```yaml
# .github/workflows/deploy.yml
name: 🚀 Deploy
on:
  pull_request: {}
 
jobs:
  test:
     # ... test steps
  deploy:
    name: Deploy
    needs:
      - test
    steps:
      # ... setup

      # This step might fail, causing the next step to never be executed
      - name: 🛫 Preview Deploy
        id: preview_deploy
        if: github.event_name == 'pull_request'
        run:  DEPLOY_SCRIPT_HERE
      - name: 💬 Notify PR
        if: github.event_name == 'pull_request'
        uses: exercism/pr-commenter-action@v1.3.0
        with:
          github-token: ${{ github.token }}
          template-variables: |
            {
              "onUpdate": "edit",
              "create": true,
              "url": "${{ steps.preview_deploy.outputs.url }}",
              "id": "${{ github.sha }}-create"
            }
```

Then another workflow is triggered when the PR is closed and shuts down the preview

```
# .github/workflows/shutdown-preview.yml
name: 🛬 Shut Down Preview
on:
  pull_request:
    types:
      - closed

jobs:
  shutdown:
    name: Shut Down
    runs-on: ubuntu-latest
    steps:
      # ... setup ...
      - name: 🛬 Shut Down Preview
        # The shutdown script already does nothing when no preview was deployed in the first place
        run:  SHUTDOWN SCRIPT

      - name: 💬 Notify PR
        if: github.event_name == 'pull_request'
        uses: exercism/pr-commenter-action@v1.3.0
        with:
          github-token: ${{ github.token }}
          template-variables: |
            {
              "onUpdate": "edit-only",
              "shutdown": true,
              "id": "${{ github.sha }}-shutdown"
            }

```

**Comment by angelikatyborska** [comment]
Alright, I think I understand your use case now 👍 I'm going to open a PR soon. But I'm not going to reuse the `onUpdate` option for this, it would invalidate its current definition ("what should happen if a comment was already created on this PR"). I'll create a new option.

**Comment by Xiphe** [comment]
Awesome! The PR looks like it will cover my case 🎉

**Comment by angelikatyborska** [comment]
I released version 1.4.0 with this feature
